coprocesses/coprocess.go

```diff
@@ -0,0 +1,105 @@
+package coprocesses
+
+import (
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/joyent/containerpilot/utils"
+)
+
+// Coprocess configures a process that will run alongside the main process
+type Coprocess struct {
+	Name     string      `mapstructure:"name"`
+	Command  interface{} `mapstructure:"command"`
+	Restarts int         `mapstructure:"restarts"`
+
+	Args       []string
+	cmd        *exec.Cmd
+	logWriters []io.WriteCloser
+}
+
+// NewCoprocesses parses json config into an array of Coprocesses
+func NewCoprocesses(raw []interface{}) ([]*Coprocess, error) {
+	var coprocesses []*Coprocess
+	if raw == nil {
+		return coprocesses, nil
+	}
+	var configs []*Coprocess
+	if err := utils.DecodeRaw(raw, &configs); err != nil {
+		return nil, fmt.Errorf("Coprocess configuration error: %v", err)
+	}
+	for _, t := range configs {
+		if err := parseCoprocess(t); err != nil {
+			return nil, err
+		}
+		coprocesses = append(coprocesses, t)
+	}
+	return coprocesses, nil
+}
+
+func parseCoprocess(coprocess *Coprocess) error {
+	args, err := utils.ToStringArray(coprocess.Command)
+	if err != nil {
+		return err
+	}
+	coprocess.Args = args
+	if coprocess.Args == nil || len(coprocess.Args) == 0 {
+		return fmt.Errorf("Coprocess did not provide a command")
+	}
+	cmd := utils.ArgsToCmd(coprocess.Args)
+	coprocess.cmd = cmd
+
+	if coprocess.Name == "" {
+		coprocess.Name = strings.Join(coprocess.Args, " ")
+	}
+	return nil
+}
+
+// Start runs the coprocess
+func (coprocess *Coprocess) Start() {
+	log.Debugf("coprocess[%s].Start", coprocess.Name)
+	fields := log.Fields{"process": "coprocess", "coprocess": coprocess.Name}
+	stdout := utils.NewLogWriter(fields, log.InfoLevel)
+	stderr := utils.NewLogWriter(fields, log.DebugLevel)
+	coprocess.logWriters = []io.WriteCloser{stdout, stderr}
+	defer coprocess.closeLogs()
+
+	for {
+		if coprocess.Restarts < 0 {
```

Comment:
In the first pass you mentioned that a dying coprocess kills the container - Which is probably preferable to restarting it a finite number of times. The traditional way is to treat `-1` special like "always restart" and then "0" means what it says on the tin. 

However, I hate doing that. I'd try to avoid doing the whole special values case if we can just put another config in.  I'd take a page out of the [upstart](http://upstart.ubuntu.com/cookbook/#respawn-limit) cook-book and make two configs:

- restart: true / fase
- restartLimit: 1+ | "unlimited"

Since we are doing raw configs, we can detect the type of `restartLimit` and actually get a nice self-documenting value.